Release workflow

paul-sachs opened the pull request:
This is a modified version of a local script i've been using for a few releases. Hasn't been tested as an action yet so we'll need a new release to validate.

### New workflows for release management:

#### Prepare Release Workflow:
* Added `.github/workflows/prepare-release.yml` to automate the creation of draft releases, release branches, and pull requests for new versions. Includes support for potential release/hotfix base branches in the case of a backport fix. To create a backport, we'll just need to create a branch for the target base version (release/v1.x) and set the base branch for the action.

#### Publish Release Workflow:
* Added `.github/workflows/publish-release.yml` to handle the publication of merged release pull requests, including publishing to npm, updating GitHub release notes

#### Update Fixed Issues Workflow:
* Added `.github/workflows/updated-fixed-issues.yml` to comment on fixed issues.

Commit:
A release workflow github action

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Format

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Add some cleanup and fix potential edge cases

Signed-off-by: Paul Sachs <psachs@buf.build>

paul-sachs commented:
After this is merged, I can run it through some dry runs of the first prepare stage which can be considered a noop if the branch is closed with accepting.

Commit:
Add support for alternative base ref

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Format again

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Review updates

- Moved to npm ci
- Split issue updates into a distinct action
- Removed unnecessary registry option on publish setup

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Naming files consistently

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Use npm trusted publishing

Signed-off-by: Paul Sachs <psachs@buf.build>

paul-sachs commented:
Alright, updated. After a little bit of click ops on each package, following [these instructions](https://docs.npmjs.com/trusted-publishers#configuring-trusted-publishing), we no longer need a token and we get providence and everything.

Commit:
Move permissions to single location

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Fixed a minor bug on cleanup and validated trusted publishing

Also removed update fixed issues action which wouldn't trigger
since the triggering event is a github action and github actions
prevents cascading actions.

Signed-off-by: Paul Sachs <psachs@buf.build>

paul-sachs commented:
Confirmed working in this repo: https://github.com/paul-sachs/npm-publish-test. Some minor adjustments have been made to fix cleanup and I've removed the fixed issues action since it wasn't triggering as a cascading action (github prevents cascading actions).

Commit:
Updated semver version check with official regex

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Improved set version requirements

Signed-off-by: Paul Sachs <psachs@buf.build>

paul-sachs commented on .github/workflows/prepare-release.yml:
```diff
@@ -0,0 +1,87 @@
+            echo "Expected format: MAJOR.MINOR.PATCH (e.g., 1.2.3) or with pre-release/build metadata (e.g., 1.2.3-alpha.1+build.1)"
+            exit 1
+          fi
+          echo "✓ Version '${{ inputs.version }}' is valid semver"
```
Sure. I've updated `set-workspace-version.js` to have an improved regexp and also add an error specifically if the version is invalid. Should be sufficient for now.

timostamm commented on .github/workflows/prepare-release.yml:
```diff
@@ -0,0 +1,78 @@
+      - name: Create draft release
+        run: gh release create v${{ inputs.version }} --draft --generate-notes
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Is there a reason to create a draft release? Publishing the draft won't publish, editing the draft doesn't work either, and we have to delete the draft if we abort a release. So I only see downsides to creating a draft.

timostamm commented on .github/workflows/prepare-release.yml:
```diff
@@ -0,0 +1,78 @@
+          git commit -s -m "Release ${{ inputs.version }}"
+          git push --set-upstream origin "release/prep-release-${{ inputs.version }}"
+
+      - name: Get release notes
```
Is this the reason to draft a release? Then I suggest that we create a draft, get the generated release notes, and delete the draft.

hudlow replied:
Why not kick the action off by manually creating a draft release and editing the release notes up front? The `prepare-release` action could trigger on that event, pick up the release tag, target branch, and release notes, create the release branch and PR, and delete the draft release.

It wouldn't be very different from what you have, but it would make use of the ["new release" UI](https://github.com/connectrpc/connect-query-es/releases/new) rather than the more esoteric UI to trigger a workflow.

One downside I see is that you could accidentally create a non-draft release in the UI... but this could maybe be handled by permissions or even detected and cleaned up by the `prepare-release` workflow?

paul-sachs replied:
The original idea was to leverage a review process, which is more central to this current version (everything pivots off the PR). Creating a release doesn't really have associated workflows so we'd need to do cleanup. But both would obviously work.

hudlow replied:
Yeah, totally get why you're using a PR for review, and I wasn't suggesting changing that. I was just suggesting using the draft release creation to kick off the flow which would seem marginally more GitHub-idiomatic to me, even though you still wouldn't be publishing the draft release actually created since it would just be tagged from `main` (or from a branch for a backport).

paul-sachs replied:
Ah i think i now understand what you meant. So instead of triggering via workflow run, we'd trigger on a draft release creation. Yeah, has a little more cleanup but i can see why that might be a little more natural.

paul-sachs replied:
@timostamm do you have thoughts on this? Using a draft release as the trigger feels a little more intuitive vs an arbitrary workflow.

timostamm replied:
Can draft releases trigger a workflow?

paul-sachs replied:
Hmm, not clear from the [docs](https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#release). It says that a draft release doesn't trigger a created/edited/deleted, which implies `published` but i don't know how you would trigger a published event on a draft release (i guess when it moves from draft to release). So yeah, that would probably be a blocker to using drafts.

timostamm commented on .github/workflows/add-to-project.yml:
Why rename this unrelated file? Can you revert?

paul-sachs replied:
Huh yeah weird, I'll revert this.

timostamm commented on .github/workflows/publish-release.yml:
```diff
@@ -0,0 +1,105 @@
+          echo "Deleting release branch ${{ github.event.pull_request.head.ref }}"
+          git push origin --delete ${{ github.event.pull_request.head.ref }}
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Should this be any different from other PRs that are closed?

I would prefer that we don't do this until we find out that we really need it.

timostamm reviewed (commented):
Nice. Let's get this merged here, and apply to other relevant repositories after running it for a while. 

scripts/release.js: Can you update the comment that recommends the procedure for releases? 

.github/RELEASING.md: Can you update this document for the new release process?

timostamm commented on .github/workflows/prepare-release.yml:
```diff
@@ -0,0 +1,78 @@
+        required: true
+        type: string
+      base_branch:
+        description: "Base branch for release (e.g. release/v1.x, hotfix/v1.2.x)"
```
Can you explain how this would be used?

Commit:
Revert file rename

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Update setup-node action

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Review comments addressed

- Updated releasing process notes
- Used API to get release notes to avoid draft release
- Use shared js util to get workspace version

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Fix lint and add some explanation of base_branch to release.md

Signed-off-by: Paul Sachs <psachs@buf.build>

Commit:
Remove tentantive support for hotfix releases

Signed-off-by: Paul Sachs <psachs@buf.build>

paul-sachs committed:
Merge branch 'main' into psachs/add-release-workflow

paul-sachs committed:
Merge branch 'main' into psachs/add-release-workflow

hudlow commented on scripts/release.js:
```diff
@@ -12,20 +12,16 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
```
Why are we keeping this file?

paul-sachs replied:
Not sure i understand, `npm run release` runs this script

hudlow replied:
Derp, of course — somehow I was thinking you moved `npm publish` directly to the workflow.

paul-sachs replied:
Admittedly, it is a little annoying how close those two commands are 🤷

hudlow commented on scripts/utils.js:
```diff
@@ -0,0 +1,50 @@
+// Copyright 2021-2023 The Connect Authors
```
Do we need this function to be externalized from the script if we delete `scripts/release.js`?

hudlow replied:
Disregard.
